File: Course_1-Algorithmic_Toolbox/Week-2/Excercise_Challenges/7_last_digit_of_the_sum_of_fibonacci_numbers_again/fibonacci_partial_sum.py
def fibonacci_partial_sum_fast(start, finish):
    F = [0, 1]
    PISANO = 60
    for _ in range(PISANO - 2):
        F.append((F[-1] + F[-2]) % 10)
    
    start %= PISANO
    finish %= PISANO
    if finish < start:
        finish += PISANO

    sm = 0
    for i in range(start, finish + 1):
        sm = (sm + F[i % PISANO]) % 10

    return sm % 10

def fibonacci_partial_sum_naive(from_, to):
    sum = 0

    current = 0
    next  = 1

    for i in range(to + 1):
        if i >= from_:
            sum += current

        current, next = next, current + next

    return sum % 10


if __name__ == '__main__':
    start, finish = map(int, input().split())
    print(fibonacci_partial_sum_fast(start, finish))

File: Course_1-Algorithmic_Toolbox/Week-2/Excercise_Challenges/7_last_digit_of_the_sum_of_fibonacci_numbers_again/fibonacci_partial_sum.cpp
#include <iostream>
#include <vector>
#define PISANO 60
typedef long long ll;
using namespace std;

int F[PISANO];

void fibonacci_series(void)
{
    F[0] = 0;
    F[1] = 1;
    for(int i = 2; i < PISANO; ++i)
        F[i] = (F[i - 1] + F[i - 2]) % 10;
}

int get_fibonacci_partial_sum_fast(ll from, ll to)
{
    from %= PISANO;
    to %= PISANO;
    if(to < from)
        to += PISANO;

    int sum = 0;
    for(int i = from; i <= to; ++i)
        sum = (sum + F[i % PISANO]) % 10;

    return sum % 10;
}

long long get_fibonacci_partial_sum_naive(long long from, long long to)
{
    long long sum = 0;

    long long current = 0;
    long long next  = 1;

    for (long long i = 0; i <= to; ++i) {
        if (i >= from) {
            sum += current;
        }

        long long new_current = next;
        next = next + current;
        current = new_current;
    }

    return sum % 10;
}

int main(void)
{
    fibonacci_series();
    ll from, to;
    cin >> from >> to;
    cout << get_fibonacci_partial_sum_fast(from, to) << '\n';

    return 0;
}
